### DISCUSSION TITLE
GitHub repository as pack source.

### DISCUSSION THREAD

**Discussion by DavidJurajdaNXP** [Open-CMSIS-Pack/Open-CMSIS-Pack-Spec#278] [Ideas]
Pack installation from public index is fine for consumption of already developed pack. On the other hand development of new pack would be managed in some repository. It would be good to have native method (beside standard pack installation) to identify such repository as pack source. It might be done in form of local index. Installation from remote repository address might be also useful.

**Comment by ReinhardKeil** [comment]
This method exists already. It is the [repository element](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/element_repository.html) - it is shown in the uVision Pack Installer and you can then use the [local pack workflow](
https://github.com/Open-CMSIS-Pack/cmsis-toolbox/blob/main/docs/build-tools.md#install-a-repository). 

The question might be: can the [repository element](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/element_repository.html) also be shown using `cpackget` or any other means.  IMHO the best way would be to generate a [pack datasheet](https://github.com/Open-CMSIS-Pack/devtools/issues/1189) as proposed a while back ago.

**Comment by DavidJurajdaNXP** [comment]
related issue: #279 Provide full, readable and editable "condition" element directly in component and file element

As @jkrech previously mentioned:
- Any *.pdsc file on disk can be registered to the ./Local/local_repository.pidx using "cpackget add <path>/<vendor>.<name>.pdsc" 
- The requirement is that the file references of the pdsc need to resolve with the checked out repository structure! 
    - @DavidJurajdaNXP  comment: we are expecting 1:1 repository/pack layout.
- Space for improvement is in automation of manual steps.

@jkrech proposal:
$cpackget checkout packID@version --destination=/my/local/packs/source
This would clone the pack respository captured for packID into /my/local/packs/source/ directory and checkout the Tag corresponding to version. It would add the pdsc file to the local repository for /my/local/packs/source/packID/packID.pdsc 
Is this what you have in mind?

@DavidJurajdaNXP  comment:
Based on question from @jkrech: 
"during development" mean "before publishing the pack via the pack index or in preparation for an update"

I think https://github.com/Open-CMSIS-Pack/cmsis-toolbox/blob/main/docs/build-tools.md#install-a-repository cover my use case.
If user modify *.pdsc or repository content is it necessary to somehow reload repository?

**Comment by jkrech** [comment]
Question from @DavidJurajdaNXP : If user modify *.pdsc or repository content is it necessary to somehow reload repository?
Answer: If your environment is an IDE which keeps the information of the loaded pdsc files "in memory", then you need to run a "reload" command in the IDE to have changes in the pdsc file reflected.
If you change the content of a file in the pack the build is expected to pick up the latest modifications. Due to the time stamp of the file CMake should identify the change and trigger the required tool invocations.

If you are running `csolution`, with each invocation the installed packs (including `local_repository.pidx` registered packs) are read and therefore any changes will be immediately reflected.

**Comment by DavidJurajdaNXP** [comment]
Local repository can be used as pack via cpackget add /..pdsc. Direct usage of remote repository is not covered and may require further discussion.
